bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java

```diff
@@ -381,25 +382,18 @@ private boolean smartSizesSupported() {
      */
 
     @Override
-    protected @NotNull ComponentData getComponentData() {
-        return new ImageDataImpl(this, resource);
-    }
-
-    @Override
-    public Resource getDataLayerAssetResource() {
-        return Optional.ofNullable(this.fileReference)
-            .map(reference -> request.getResourceResolver().getResource(reference))
-            .orElse(null);
-    }
-
-    @Override
-    public String getDataLayerTitle() {
-        return title;
-    }
-
-    @Override
-    public String getDataLayerLinkUrl() {
-        return getLink();
+    @NotNull
+    protected ImageData getComponentData() {
+        return DataLayerBuilder.extending(super.getComponentData()).asImageComponent()
+            .withTitle(this::getTitle)
+            .withLinkUrl(this::getLink)
+            .withAssetData(() ->
```

Comment:
What if the `Supplier<AssetData>` would be just an anonymous function that returns a locally defined `AssetData` instance?

Reply:
Or just use an `AssetData` as parameter of `withAssetData`?

Reply:
My thought here was that if you already have an `Asset` then you can use this to get the AssetData for that asset.
```JAVA
AssetData data = DataLayerBuilder.forAsset(Asset).build()
```
This is just a typical implementation, you override any of the suppliers before doing the `build()` if there is some way that your model should differ from what this helper provides.

If you do not have an Asset then you can use this
```JAVA
AssetData data = DataLayerBuilder.forAsset()
  .withId(/**/)
  ...
  .build()
```

Regardless of how you get the AssetData, including it in the component model will be like this:
```JAVA
ImageData imageData = DataLayerBuilder.forImage()
  .withAssetData(() -> assetData)
  .build();
```

This pattern encourages you to use the builder for creating AssetData, but if you really wanted to you could create some class that implements `AssetData` and then supply an instance of that to the `.withAssetData(() -> customAssetDataInstance)`. That said, i don't think there is any reason to do this since the builder can create the data you want.

Reply:
That's exactly what I was suggesting: creating `AssetData` once and mapping it directly in `withAssetData` instead of mapping a function that recreates it on every call of `getAssetData`.

Reply:
Hello,

I agree with you that I don't like the amount of re-work that is done if the `ComponentDataImpl#getAssetData()` method is called twice.

### Proposal 1
The computation of the `AssetDataImpl` object could be moved outside of the supplier, and cached similar to how other getters are done in models - such that the block is only executed once.

```JAVA
    private AssetData getAssetData() {
        if (this.assetData == null) {
            this.assetData = Optional.ofNullable(this.fileReference)
                .map(reference -> this.request.getResourceResolver().getResource(reference))
                .map(assetResource -> assetResource.adaptTo(Asset.class))
                .map(DataLayerBuilder::forAsset)
                .map(AssetDataBuilder::build)
                .orElse(null);
        }
        return this.assetData;
    }

    @Override
    @NotNull
    protected ImageData getComponentData() {
        return DataLayerBuilder.extending(super.getComponentData()).asImageComponent()
            .withTitle(this::getTitle)
            .withLinkUrl(this::getLink)
            .withAssetData(this::getAssetData)
            .build();
    }
```

However, this isn't fully optimal because, while it does eliminate (potentially) duplicate

- resolution of the asset resource
- adaptation of that resource to an Asset object
- creation/building of the data layer builder (not very expensive since no heavy lifting is done)

none of the data builder methods are caching. 
In the above solution, repeated calls to `ComponentDataImpl#getAssetData()` would always return the same object; however, subsequent repeated calls to the returned object getters (e.g. `AssetData#get*()`) would still re-incur the work of calculating the values. It may be worth nothing that the current [`AssetDataImpl`](https://github.com/adobe/aem-core-wcm-components/blob/2f24bf5/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/AssetDataImpl.java) does not cache the values of the getters and recomputes them on-demand, so the above change would result in the same behaviour as currently exists. 

As it stands now (in this PR), the builder does not produce an object which actually contains the computed values, it is more of just setting up the wiring to know how to get the values when needed. In this way the creation of the component/asset data is fairly trivial, the hard work is only done on-demand.

---
### Proposal 2
A solution that would eliminate this issue across the board, in its entirety, would be to cache the values produced by the suppliers inside the `AssetDataImpl` and `ComponentDataImpl`. Then the documentation could make a simple statement that the values are not computed until they are requested, but once requested are never re-computed.

This would change the `ComponentDataImpl#getAssetData()` method from:
```JAVA
    @Override
    @Nullable
    public AssetData getAssetData() {
        return this.getDataLayerSupplier()
            .getAssetData()
            .map(Supplier::get)
            .orElse(null);
    }
```


TO
```JAVA
    @Override
    @Nullable
    public AssetData getAssetData() {
        if (this.assetData == null) {
            this.assetData = this.getDataLayerSupplier()
                .getAssetData()
                .map(Supplier::get)
                .orElse(null);
        }
        return this.assetData;
    }
```

Similar updates would then be done on all the other `Asset/ComponentDataImpl#get*()` methods, which would entirely eliminate the need to worry about the caching issue. 

This isn't as technically complex as the original solution of allowing the caching strategy to be set, instead it's an always cache strategy. 

I favor this solution because it solves the issue but allows for a consistent expectation across the entire data builder, regardless of what is being built. I also think it most likely matches the expectations of the user (developer).

---
### Proposal 3

There are other options, and I think that this might be more in line with what you are thinking.
The other option is the restore the original [`AssetDataImpl`](https://github.com/adobe/aem-core-wcm-components/blob/2f24bf5/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/AssetDataImpl.java) that can be constructed from an `Asset`; and then there would be two `AssetData` implementations (one backed by the builder suppliers, the other is just an object)

In this solution, the `DataLayerBuilder#forAsset(Asset)` method would be changed thusly:
From:
```JAVA
    public static AssetDataBuilder forAsset(@NotNull final Asset asset) {
         return DataLayerBuilder.forAsset()
            .withId(asset::getID)
            ...;
    }
```
To:
```JAVA
    public static AssetData forAsset(@NotNull final Asset asset) {
        new AssetDataImpl(asset);
    }
```

Note that the `DataBuilder#forAsset(Asset)` no longer returns a builder, it returns the asset data object directly.
Any extension or override of that would have to be done by
```JAVA
    DataLayerBuilder.extending(DataLayerBuilder.forAsset(asset))
    ...
    .build()
```

but anything provided after extending the asset data would, again, be lazy and re-computed on every call. 
Also note that the original [`AssetDataImpl`](https://github.com/adobe/aem-core-wcm-components/blob/2f24bf5/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/AssetDataImpl.java) recomputes it's values on every call. As far as I can tell, this solution doesn provide any benefits over **Solution 1** unless we also add caching to [`AssetDataImpl`](https://github.com/adobe/aem-core-wcm-components/blob/2f24bf5/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/AssetDataImpl.java), which would have the same effect as adding it to the supplier backed `AssetDataImpl` provided in this branch (effectively **Solution 2** (_bold just to match title, not for emphasis_)).

---

p.s.: Thank you for taking the time to review this in depth and find these issues; I appreciate the feedback.

Reply:
My preference goes to your proposal#2, which goes more towards what we've tried to achieve in the model implementations.
I would not only cache the return value of `ComponentDataImpl.getAssetData()`method but of all methods of `ComponentDataImpl` and `AssetDataImpl`.

Reply:
Sorry for the late answered, I hoped I would get a chance to checkout the changes and maybe come up with another proposal. *Solution 2* is my favourite too.

Reply:
Thank you, will implement.
For future reference, if the Guava dependency is ever updated to a newer version, there is a MemoizingSupplier class in there that would resolve this issue cleanly, and would also ensure a single execution (i.e. don't try again if it returned null).